src/zenml/integrations/kubernetes/orchestrators/kubernetes_orchestrator.py

```diff
@@ -0,0 +1,436 @@
+            stack (Stack): ZenML stack.
+            runtime_configuration (RuntimeConfiguration): _description_
+
+        Raises:
+            RuntimeError: if trying to run from a Jupyter notebook.
+        """
+
+        # First check whether the code running in a notebook
+        if Environment.in_notebook():
+            raise RuntimeError(
+                "The Kubernetes orchestrator cannot run pipelines in a notebook "
+                "environment. The reason is that it is non-trivial to create "
+                "a Docker image of a notebook. Please consider refactoring "
+                "your notebook cells into separate scripts in a Python module "
+                "and run the code outside of a notebook when using this "
+                "orchestrator."
+            )
+
+        assert runtime_configuration.run_name, "Run name must be set"
+
+        run_name = runtime_configuration.run_name
+        pipeline_name = pipeline.name
+
+        # Get docker image name (for all pods).
+        image_name = self.get_docker_image_name(pipeline.name)
+        image_name = get_image_digest(image_name) or image_name
+
+        # Get step dependencies, e.g., {"step": ["upstream_step_1", ...], ...}
+        step_dependencies = {
+            step.name: self.get_upstream_step_names(step, pb2_pipeline)
+            for step in sorted_steps
+        }
+
+        # Build entrypoint command and args for the orchestrator pod.
+        # This will internally also build the command/args for all step pods.
+        command = (
+            KubernetesOrchestratorEntrypointConfiguration.get_entrypoint_command()
+        )
+        args = KubernetesOrchestratorEntrypointConfiguration.get_entrypoint_arguments(
+            run_name=run_name,
+            pipeline_name=pipeline_name,
+            image_name=image_name,
+            kubernetes_namespace=self.kubernetes_namespace,
+            pb2_pipeline=pb2_pipeline,
+            sorted_steps=sorted_steps,
+            step_dependencies=step_dependencies,
+        )
+
+        # Build manifest for the orchestrator pod.
+        pod_name = tfx_kube_utils.sanitize_pod_name(run_name)
+        pod_manifest = build_base_pod_manifest(
+            run_name=run_name,
+            pipeline_name=pipeline_name,
+            image_name=image_name,
+        )
+        pod_manifest = update_pod_manifest(
+            base_pod_manifest=pod_manifest,
+            pod_name=pod_name,
+            command=command,
+            args=args,
+        )
+        pod_manifest["spec"]["serviceAccountName"] = "zenml-service-account"
+
+        # Create and run the orchestrator pod.
+        core_api = tfx_kube_utils.make_core_v1_api()
+        core_api.create_namespaced_pod(
+            namespace=self.kubernetes_namespace,
+            body=pod_manifest,
+        )
+        logger.info("Kubernetes orchestrator pod started.")
+
+        # Wait for the orchestrator pod to finish.
+        tfx_kube_utils.wait_pod(
```

Comment:
This seems to work very differently from our Kubeflow orchestrator, does the kubernetes orchestrator not build a dag locally and instead the master node runs steps when they become ready?

Reply:
And how does the wait_pod interact with an orchestrator configuration with synchronous=False

Reply:
- We also build a DAG locally first, which is captured in the `step_dependencies` variable (which should probably be renamed as discussed above). We then start the master node, which receives the DAG, and spins up pods for all steps that have no previous steps. After a step is completed, the master node then checks for all other steps whether all their respective previous steps were run, and, if so, also spins up those pods.
- Great point, synchronous=False is not working currently.

Reply:
Implemented snchronous=True.